sympy/core/compatibility.py

```diff
@@ -676,6 +677,36 @@ def ordered(seq, keys=None, default=True, warn=False):
         yield from d[k]
         d.pop(k)
 
+class OpDispatcher(Dispatcher):
+    """
+    Dispatcher for non-associative operations, e.g. pow
+
+    .. notes::
+       This approach is experimental, and can be replaced or deleted in the future.
+       See https://github.com/sympy/sympy/pull/19463.
+
+    See Also
+    ========
+
+    core.operations.AssocOpDispatcher
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.fallback = fallback
+        self.handlermethod = "_%s_handler" % name
+        super().__init__(name, doc)
+
+    def __call__(self, *args, **kwargs):
+        try:
+            return super().__call__(*args, **kwargs)
+        except NotImplementedError:
+            return self.fallback(*args, **kwargs)
+
+    def add(self, signature, func, on_ambiguity=(lambda *args, **kwargs: None)):
+        # This suppresses unnecessary ambiguity warning, since not all sympy objects
+        # need to operate with each other.
```

Comment:
Can you elaborate on this? Perhaps with an example that would warn that you don't want to warn?

Reply:
Suppose you want to register `Vector` and `AccumBounds` to the dispatcher, who are clearly not meant to be added or multiplied with each other. You would register like this:
```python
>>> @add.register(Expr, Accumbounds)
... def _(a, b):
...     ...
>>> @add.register(Expr, Vector)
... def _(a, b):
...     ...
```
But this results ambiguity warning, saying that you have to register `@add.register(Vector, AccumBounds)` as well since these two are both subclass of `Expr`.  This code suppresses it because you don't want to register every combination of classes that are not meant to be compatible.

Reply:
Your example doesn't fail:
```
import warnings
warnings.simplefilter("error")
from sympy import Expr, AccumBounds
from sympy.physics.vector import Vector
from sympy.multipledispatch import Dispatcher
add = Dispatcher("add")
@add.register(Expr, AccumBounds)
def _(a, b):
    ...
@add.register(Expr, Vector)
def _(a, b):
    ...
```
Can you produce one which does?

Reply:
Oh, I oversimplified my example. As depicted in doctest of `AssocOpDispatcher`, each class must be registered thrice: e.g. `(Expr, Vector)`, `(Vector, Expr)` and `(Vector, Vector)`
This is what happens when I change the code to disable `on_ambiguity` suppression:

```python
>>> from sympy import Expr, Add
>>> class NewExpr(Expr):
...     pass

>>> @add.register_priority(Expr, NewExpr)
... def _(a,b):
...     pass

/home/codes/sympy/sympy/sympy/multipledispatch/dispatcher.py:28: AmbiguityWarning: 
Ambiguities exist in dispatched function add

The following signatures may result in ambiguous behavior:
        [Expr, NewExpr], [MatrixExpr, Expr]


Consider making the following additions:

@dispatch(MatrixExpr, NewExpr)
def add(...)
  warn(warning_text(dispatcher.name, ambiguities), AmbiguityWarning)
```

It requires `MatrixExpr` and `NewExpr` to be registered.

Reply:
Right, that warning is real I think. If i call with `MatrixExpr, NewExpr`, which dispatcher do you want to be called? The warning is telling you that sympy will arbitrarily pick between the other two overloads, because _both_ match.

> you don't want to register every combination of classes that are not meant to be compatible.

Are you sure you don't want to explicitly register them to return `NotImplemented` or similar?

Reply:
If `MatrixExpr` and `NewExpr` need to be compatible with each other, then obviously these two should be explicitly registered. If they are not meant to be compatible, then we couldn't care less about what happens when we add these two.  Explicitly registering every possible combination will be overwhelming and unnecessary.

Reply:
> If they are not meant to be compatible, then we couldn't care less about what happens when we add these two.

I disagree with that. If they're not meant to work together, an error should be raised somewhere.

> Explicitly registering every possible combination will be overwhelming

Can we implicitly register extra handlers for these cases in the ambiguity callback?